### PR TITLE
feat: route-level MTLS

### PR DESCRIPTION
**Pull request by kingluo**
### Description

https://lists.apache.org/thread/t8kmt7hf0krvf2mc0cf0zp5q330dmo1d

### Checklist

- [x] I have explained the need for this PR and the problem it solves
- [x] I have explained the changes or the new features added to this PR
- [x] I have added tests corresponding to this change
- [x] I have updated the documentation to reflect this change
- [x] I have verified that this change is backward compatible (If not, please discuss on the [APISIX mailing list](https://github.com/apache/apisix/tree/master#community) first)

### CODE REVIEW

File: docs/en/latest/tutorials/client-to-apisix-mtls.md [hunk 1]
```diff
@@ -193,6 +193,127 @@ curl --resolve "test.com:9443:127.0.0.1" https://test.com:9443/anything -k --cer
...
+## MTLS bypass based on regular expression matching against URI
+
+APISIX allows configuring an URI whitelist to bypass MTLS.
+If the URI of a request is in the whitelist, then the client certificate will not be checked.
+Note that other URIs of the associated SNI will get HTTP 400 response
+instead of alert error in the SSL handshake phase, if the client certificate is missing or invalid.
+
+### Timing diagram
+
+![skip mtls](https://raw.githubusercontent.com/apache/apisix/master/docs/assets/images/skip-mtls.png)
```
Comment (An-DJ) [inline docs/en/latest/tutorials/client-to-apisix-mtls.md:205]: Better to use relative path as below:

https://github.com/apache/apisix/blob/master/docs/en/latest/benchmark.md?plain=1#L40

File: docs/zh/latest/tutorials/client-to-apisix-mtls.md [hunk 2]
```diff
@@ -193,6 +193,122 @@ curl --resolve "test.com:9443:127.0.0.1" https://test.com:9443/anything -k --cer
 
 由于我们在示例中配置了 `proxy-rewrite` 插件，我们可以看到响应体中包含上游收到的请求体，包含了正确数据。
 
+## 基于对 URI 正则表达式匹配，绕过 MTLS
+
+APISIX 允许配置 URI 白名单以便绕过 MTLS。如果请求的 URI 在白名单内，客户端证书将不被检查。注意，如果针对白名单外的 URI 发请求，而该请求缺乏客户端证书或者提供了非法客户端证书，会得到 HTTP 400 响应，而不是在 SSL 握手阶段被拒绝。
+
+### 时序图
+
+![skip mtls](https://raw.githubusercontent.com/apache/apisix/master/docs/assets/images/skip-mtls.png)
```
Comment (An-DJ) [inline docs/zh/latest/tutorials/client-to-apisix-mtls.md:202]: ditto

File: apisix/init.lua [hunk 3]
```diff
@@ -179,6 +180,8 @@ function _M.http_ssl_phase()
 
     local ok, err = router.router_ssl.match_and_set(api_ctx)
```
Comment (monkeyDluffy6017) [inline apisix/init.lua:181]: if reject, is ok equal to `false` ?
if ok equal to `false`, does `ngx_exit(-1)` interrupt the execution of the current request ?

File: apisix/init.lua [hunk 4]
```diff
@@ -310,12 +313,38 @@ local function verify_tls_client(ctx)
...
+        local res = ctx.var.ssl_client_verify
+        if res ~= "SUCCESS" then
+            if res == "NONE" then
+                core.log.error("client certificate was not present")
+            else
+                core.log.error("client certificate verification is not passed: ", res)
+            end
+
+            return false
+        end
```
Comment (monkeyDluffy6017) [inline apisix/init.lua:345]: This code can be wrapped into a function

File: docs/en/latest/tutorials/client-to-apisix-mtls.md [hunk 5]
```diff
@@ -193,6 +193,127 @@ curl --resolve "test.com:9443:127.0.0.1" https://test.com:9443/anything -k --cer
...
+
+### Example
+
+```bash
+curl http://127.0.0.1:9180/apisix/admin/routes/1 \
```
Comment (monkeyDluffy6017) [inline docs/en/latest/tutorials/client-to-apisix-mtls.md:210]: Maybe we should use three blocks, don't wrap them into one code block
1. config the route and ssl
2. if the client certificate is missing and the URI is not in the whitelis
3. the client certificate is missing, but the URI is in the whitelist

File: apisix/init.lua (hunk 4 above)
Comment (kingluo) [reply apisix/init.lua:345]: Not necessary, only logging statements there. Not so much intersection.
The original code is for host check, which is a different logic, and it's better to keep the original version.

File: apisix/init.lua (hunk 3 above)
Comment (kingluo) [reply apisix/init.lua:181]: It's the original logic and irrelevant to this PR.
On the contrary, the route-level MTLS bypasses the client cert error in the SSL phase, i.e. it relaxes the success restrictions.

File: docs/en/latest/tutorials/client-to-apisix-mtls.md (hunk 5 above)
Comment (kingluo) [reply docs/en/latest/tutorials/client-to-apisix-mtls.md:212]: done

File: docs/zh/latest/tutorials/client-to-apisix-mtls.md [hunk 6]
```diff
@@ -193,6 +193,122 @@ curl --resolve "test.com:9443:127.0.0.1" https://test.com:9443/anything -k --cer
...
+
+### 例子
+
+```bash
```
[the remark this answers is not in the document]
Comment (kingluo) [reply docs/zh/latest/tutorials/client-to-apisix-mtls.md:208]: done

**Review by monkeyDluffy6017 — APPROVED** [review]

**Review by soulbird — APPROVED** [review]

**Review by An-DJ — APPROVED** [review]

**Pull request merged**